Printing waveformInfo type fails with AttributeError

* Printing waveformInfo type fails with AttributeError

* 
 capture a waveform on the scope and print(waveform)


self = <[AttributeError("'WaveformInfo' object has no attribute 'actual_samples'") raised in repr()] WaveformInfo object at 0x2592e92c4e0>

    def __repr__(self):
        return '{0}(absolute_initial_x={1}, relative_initial_x={2}, x_increment={3}, actual_samples={4}, offset={5}, gain={6})'.format(
            self.__class__.__name__, self.absolute_initial_x, self.relative_initial_x, self.x_increment,
>           self.actual_samples, self.offset, self.gain)
E       AttributeError: 'WaveformInfo' object has no attribute 'actual_samples'

C:\Program Files\Python36\lib\site-packages\niscope\waveform_info.py:72: AttributeError

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,10 +17,11 @@ All notable changes to this project will be documented in this file.
 ## [Unreleased]
 * ### ALL
     * #### Added
-        * Support for Python 3.7
-        * \_\_version\_\_ for all drivers
+        * Support for Python 3.7 - [#895](https://github.com/ni/nimi-python/issues/895)
+        * \_\_version\_\_ for all drivers - [#928](https://github.com/ni/nimi-python/issues/928)
     * #### Changed
         * No longer globally set warnings filter for `DriverWarning` - if you want all warnings from the driver, you will need to set `warnings.filterwarnings("always", category=<driver>.DriverWarning)` in your code
+        * Fix \_\_repr\_\_ for niscope.WaveformInfo - [#920](https://github.com/ni/nimi-python/issues/920)
     * #### Removed
 * ### NI-DMM
     * #### Added

--- a/generated/niscope/waveform_info.py
+++ b/generated/niscope/waveform_info.py
@@ -69,7 +69,7 @@ class WaveformInfo(object):
     def __repr__(self):
         return '{0}(absolute_initial_x={1}, relative_initial_x={2}, x_increment={3}, actual_samples={4}, offset={5}, gain={6})'.format(
             self.__class__.__name__, self.absolute_initial_x, self.relative_initial_x, self.x_increment,
-            self.actual_samples, self.offset, self.gain)
+            self.actual_samples if hasattr(self, 'actual_samples') else 0, self.offset, self.gain)
 
     def __str__(self):
         row_format_g = '{:<20}: {:,.6g}'

--- a/generated/niscope/waveform_info.py
+++ b/generated/niscope/waveform_info.py
@@ -72,26 +72,26 @@ class WaveformInfo(object):
             self.actual_samples if hasattr(self, 'actual_samples') else 0, self.offset, self.gain)
 
     def __str__(self):
-        row_format_g = '{:<20}: {:,.6g}'
-        row_format_d = '{:<20}: {:,}'
-        row_format_s = '{:<20}: {:}'
+        row_format_g = '{:<20}: {:,.6g}\n'
+        row_format_d = '{:<20}: {:,}\n'
+        row_format_s = '{:<20}: {:}\n'
         string_representation = ''
         try:
-            string_representation += row_format_s.format('channel', self.channel) + '\n'
-            string_representation += row_format_d.format('record', self.record) + '\n'
+            string_representation += row_format_s.format('channel', self.channel)
+            string_representation += row_format_d.format('record', self.record)
         except AttributeError:
             pass
-        string_representation += row_format_g.format('Absolute X0', self.absolute_initial_x) + '\n'
-        string_representation += row_format_g.format('Relative X0', self.relative_initial_x) + '\n'
-        string_representation += row_format_g.format('dt', self.x_increment) + '\n'
+        string_representation += row_format_g.format('Absolute X0', self.absolute_initial_x)
+        string_representation += row_format_g.format('Relative X0', self.relative_initial_x)
+        string_representation += row_format_g.format('dt', self.x_increment)
         try:
-            string_representation += row_format_d.format('actual samples', self.actual_samples) + '\n'
+            string_representation += row_format_d.format('actual samples', self.actual_samples)
         except AttributeError:
             pass
-        string_representation += row_format_g.format('offset', self.offset) + '\n'
-        string_representation += row_format_g.format('gain', self.gain) + '\n'
+        string_representation += row_format_g.format('offset', self.offset)
+        string_representation += row_format_g.format('gain', self.gain)
         try:
-            string_representation += row_format_g.format('wfm length', len(self.samples)) + '\n'
+            string_representation += row_format_g.format('wfm length', len(self.samples))
         except AttributeError:
             pass
         return string_representation

--- a/src/niscope/custom_types/waveform_info.py
+++ b/src/niscope/custom_types/waveform_info.py
@@ -69,7 +69,7 @@ class WaveformInfo(object):
     def __repr__(self):
         return '{0}(absolute_initial_x={1}, relative_initial_x={2}, x_increment={3}, actual_samples={4}, offset={5}, gain={6})'.format(
             self.__class__.__name__, self.absolute_initial_x, self.relative_initial_x, self.x_increment,
-            self.actual_samples, self.offset, self.gain)
+            self.actual_samples if hasattr(self, 'actual_samples') else 0, self.offset, self.gain)
 
     def __str__(self):
         row_format_g = '{:<20}: {:,.6g}\n'

--- a/src/niscope/custom_types/waveform_info.py
+++ b/src/niscope/custom_types/waveform_info.py
@@ -72,26 +72,26 @@ class WaveformInfo(object):
             self.actual_samples, self.offset, self.gain)
 
     def __str__(self):
-        row_format_g = '{:<20}: {:,.6g}'
-        row_format_d = '{:<20}: {:,}'
-        row_format_s = '{:<20}: {:}'
+        row_format_g = '{:<20}: {:,.6g}\n'
+        row_format_d = '{:<20}: {:,}\n'
+        row_format_s = '{:<20}: {:}\n'
         string_representation = ''
         try:
-            string_representation += row_format_s.format('channel', self.channel) + '\n'
-            string_representation += row_format_d.format('record', self.record) + '\n'
+            string_representation += row_format_s.format('channel', self.channel)
+            string_representation += row_format_d.format('record', self.record)
         except AttributeError:
             pass
-        string_representation += row_format_g.format('Absolute X0', self.absolute_initial_x) + '\n'
-        string_representation += row_format_g.format('Relative X0', self.relative_initial_x) + '\n'
-        string_representation += row_format_g.format('dt', self.x_increment) + '\n'
+        string_representation += row_format_g.format('Absolute X0', self.absolute_initial_x)
+        string_representation += row_format_g.format('Relative X0', self.relative_initial_x)
+        string_representation += row_format_g.format('dt', self.x_increment)
         try:
-            string_representation += row_format_d.format('actual samples', self.actual_samples) + '\n'
+            string_representation += row_format_d.format('actual samples', self.actual_samples)
         except AttributeError:
             pass
-        string_representation += row_format_g.format('offset', self.offset) + '\n'
-        string_representation += row_format_g.format('gain', self.gain) + '\n'
+        string_representation += row_format_g.format('offset', self.offset)
+        string_representation += row_format_g.format('gain', self.gain)
         try:
-            string_representation += row_format_g.format('wfm length', len(self.samples)) + '\n'
+            string_representation += row_format_g.format('wfm length', len(self.samples))
         except AttributeError:
             pass
         return string_representation

